geode-redis/src/main/java/org/apache/geode/redis/internal/executor/set/DeltaSet.java

```diff
@@ -0,0 +1,261 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.executor.set;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.geode.DataSerializable;
+import org.apache.geode.DataSerializer;
+import org.apache.geode.Delta;
+import org.apache.geode.InvalidDeltaException;
+import org.apache.geode.cache.Region;
+import org.apache.geode.redis.internal.ByteArrayWrapper;
+
+/**
+ * TODO: it is probably a bad idea for this class to implement Set.
+ * We want to be careful how other code interacts with these instances
+ * to make sure that no modifications are made that are not thread safe
+ * and that will always be stored in the region.
+ * Currently the only "correct" methods on this class are:
+ * members, delete, customAddAll, customRemoveAll, and the
+ * serialization methods.
+ */
+public class DeltaSet implements Delta, DataSerializable {
+
+  public static long sadd(Region<ByteArrayWrapper, DeltaSet> region,
+      ByteArrayWrapper key,
+      ArrayList<ByteArrayWrapper> membersToAdd) {
+    while (true) {
```

Comment:
If we synchronize "lower" in the stack, can we avoid the complexity of looping and throwing exceptions. Can we somehow guarantee that on a given key, only one of these operations is happening simultaneously? Maybe an exclusive lock on the key?

Reply:
We could even use some kind of custom read-write lock...sadds should be able to happen in parallel, and srems should be able to happen in parallel exclusive of sadds, and dels just have to be exclusive of everything.